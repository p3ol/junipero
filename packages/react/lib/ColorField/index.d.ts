import {
  Props as DismissProps,
} from '@floating-ui/react/src/hooks/useDismiss';
import { ReactNode, MutableRefObject } from 'react';

export declare type ColorFieldRef = {
  dirty: boolean;
  isJunipero: boolean;
  valid: boolean;
  value: string;
  blur(): void;
  close(): void;
  focus(): void;
  open(): void;
  reset(): void;
  toggle(): void;
  colorAlphaRef: MutableRefObject<any>;
  colorHueRef: MutableRefObject<any>;
  colorLightnessRef: MutableRefObject<any>;
  innerRef: MutableRefObject<any>;
  textFieldRef: MutableRefObject<any>;
};

declare interface ColorFieldProps extends React.ComponentPropsWithRef<any> {
  autoFocus?: boolean;
  className?: string;
  disabled?: boolean;
  dismissOptions?: DismissProps;
  format?: 'auto' | 'hex' | 'rgb' | 'rgba' | 'hsla';
  globalEventsTarget: EventTarget;
  id?: string;
  name?: string;
  placeholder?: string;
  required?: boolean;
  tabIndex?: number;
  trigger?: 'click' | 'hover' | 'manual' | 'focus';
  valid?: boolean;
  value?: string;
  animateMenu(
    menu: ReactNode | JSX.Element,
    opts: { opened: boolean }
  ): ReactNode | JSX.Element;
  onBlur?(e: Event): void;
  onChange?(props: { value: string; valid: boolean }): void;
  onFocus?(e: Event): void;
  onToggle?({ opened }: { opened: boolean }): void;
  onValidate?(
    value: string,
    { dirty, required }: { dirty?: boolean; required?: boolean }
  ): boolean;
  ref?: MutableRefObject<ColorFieldRef | undefined>;
}

declare function ColorField(props: ColorFieldProps): ReactNode | JSX.Element;

export default ColorField;
