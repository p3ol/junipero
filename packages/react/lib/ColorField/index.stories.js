import { action } from '@storybook/addon-actions';
import { slideInDownMenu } from '@junipero/transitions';

import FieldControl from '../FieldControl';
import Label from '../Label';
import Abstract from '../Abstract';
import ColorField from '.';

export default { title: 'react/ColorField' };

export const basic = () => (
  <ColorField placeholder="Pick a color" onChange={action('change')} />
);

export const withLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info" htmlFor="firstname">Eye color</Label>
    <ColorField
      id="firstname"
      placeholder="Pick a color"
      onChange={action('change')}
      onValidate={() => false}
    />
    <Abstract className="info">
      Take a closer look at your eyes in a mirror
    </Abstract>
  </FieldControl>
);

export const autoFocused = () => (
  <ColorField
    autoFocus={true}
    onChange={action('change')}
  />
);

export const alwaysOpened = () => (
  <ColorField autoFocus={true} trigger="manual" onChange={action('change')} />
);

export const animated = () => (
  <ColorField animateMenu={slideInDownMenu} />
);
