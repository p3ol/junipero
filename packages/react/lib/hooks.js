import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
