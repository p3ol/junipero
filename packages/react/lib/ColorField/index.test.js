import { createRef } from 'react';
import {
  render,
  fireEvent,
  act,
  waitFor,
  createEvent,
} from '@testing-library/react';

import ColorField from './';

describe('<ColorField />', () => {
  it('should render', () => {
    const { container, unmount } = render(<ColorField autoFocus={true} />);
    expect(container.querySelectorAll('.junipero.color-input').length).toBe(1);
    expect(container.querySelectorAll('.pigment').length).toBe(1);
    unmount();
  });

  it('should provide some imperative handles', () => {
    const ref = createRef();
    const { container, unmount } = render(<ColorField ref={ref} />);
    expect(ref.current.innerRef).toBeDefined();
    expect(container.querySelector('.junipero.color-input'))
      .toBe(ref.current.innerRef.current);
    expect(ref.current.inputRef).toBeDefined();
    expect(ref.current.colorWheelRef).toBeDefined();
    expect(ref.current.colorLightnessRef).toBeDefined();
    expect(ref.current.colorHueRef).toBeDefined();
    expect(ref.current.colorAlphaRef).toBeDefined();
    expect(ref.current.opened).toBe(false);
    expect(ref.current.reset).toBeDefined();
    expect(ref.current.focus).toBeDefined();
    expect(ref.current.blur).toBeDefined();
    expect(ref.current.h).toBe(0);
    expect(ref.current.s).toBe(0);
    expect(ref.current.v).toBe(0);
    expect(ref.current.a).toBe(100);
    expect(ref.current.valid).toBe(false);
    unmount();
  });

  it('should update internal value if value prop changes', () => {
    const ref = createRef();
    const { rerender, unmount } = render(<ColorField ref={ref} value="#FFF" />);
    rerender(<ColorField ref={ref} value="#000" />);
    expect(ref.current.internalValue).toBe('#000');
    unmount();
  });

  it('should fire onFocus/onBlur events', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const { container, unmount } = render(
      <ColorField onFocus={onFocus} onBlur={onBlur} />
    );
    container.querySelector('input').focus();
    expect(onFocus).toHaveBeenCalled();
    container.querySelector('input').blur();
    expect(onBlur).toHaveBeenCalled();
    unmount();
  });

  it('should update values when mouse is down and lightness handle is ' +
    'selected', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );
    fireEvent.mouseDown(container.querySelector('.lightness'), { button: 0 });
    expect(ref.current.handleMoving).toBe(true);
    expect(ref.current.handleType).toBe('lightness');

    // Man testing is weird
    await act(async () => {
      window.pageXOffset = 50;
      window.pageYOffset = 50;

      // Simulate element width/height
      Object.defineProperty(
        ref.current.colorLightnessRef.current,
        'offsetWidth',
        { value: 100 }
      );
      Object.defineProperty(
        ref.current.colorLightnessRef.current,
        'offsetHeight',
        { value: 100 }
      );
    });

    // Testing library (or jsdom?) doesn't support pageX/Y if passed
    // using fireEvent.mouseMove (lol), so we have to first create the event,
    // hard code the pageX and pageY values using .defineProperty, and then
    // give them straight to fireEvent (not fireEvent.mouseMove, this does
    // not work either)
    const moveEvent = createEvent
      .mouseMove(container.querySelector('.lightness'));
    Object.defineProperty(moveEvent, 'pageX', { value: 100 });
    Object.defineProperty(moveEvent, 'pageY', { value: 100 });

    fireEvent(container.querySelector('.lightness'), moveEvent);
    expect(ref.current.s).toBe(50);
    expect(ref.current.v).toBe(50);
    unmount();
  });

  it('should update values when mouse is down and hue handle is ' +
    'selected', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );

    fireEvent.mouseDown(container.querySelector('.hue'), { button: 0 });
    expect(ref.current.handleMoving).toBe(true);
    expect(ref.current.handleType).toBe('hue');

    await act(async () => {
      window.pageXOffset = 50;
      Object.defineProperty(
        ref.current.colorHueRef.current,
        'offsetWidth',
        { value: 100 }
      );
    });

    const moveEvent = createEvent
      .mouseMove(container.querySelector('.hue'));
    Object.defineProperty(moveEvent, 'pageX', { value: 100 });
    fireEvent(container.querySelector('.hue'), moveEvent);

    expect(ref.current.h).toBe(180);
    unmount();
  });

  it('should update values when mouse is down and alpha handle is ' +
    'selected', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );

    fireEvent.mouseDown(container.querySelector('.alpha'), { button: 0 });
    expect(ref.current.handleMoving).toBe(true);
    expect(ref.current.handleType).toBe('alpha');

    await act(async () => {
      window.pageXOffset = 50;
      Object.defineProperty(
        ref.current.colorAlphaRef.current,
        'offsetWidth',
        { value: 100 }
      );
    });

    const moveEvent = createEvent
      .mouseMove(container.querySelector('.alpha'));
    Object.defineProperty(moveEvent, 'pageX', { value: 100 });
    fireEvent(container.querySelector('.alpha'), moveEvent);

    expect(ref.current.a).toBe(50);
    unmount();
  });

  it('should close color menu and trigger onBlur event if clicked ' +
    'outside', async () => {
    const ref = createRef();
    const onBlur = jest.fn();

    const { container, unmount } = render(
      <ColorField onBlur={onBlur} ref={ref} />
    );

    container.querySelector('input').focus();
    await waitFor(() => {
      expect(ref.current.opened).toBe(true);
    });

    container.querySelector('input').blur();
    fireEvent.click(document.body);
    await waitFor(() => {
      expect(ref.current.opened).toBe(false);
    });

    expect(onBlur).toHaveBeenCalled();
    unmount();
  });

  it('should not close color menu if input is clicked', async () => {
    const ref = createRef();
    const { container, unmount } = render(<ColorField ref={ref} />);
    container.querySelector('input').focus();
    await waitFor(() => {
      expect(ref.current.opened).toBe(true);
    });
    fireEvent.click(container.querySelector('.junipero.dropdown-toggle'));
    expect(ref.current.opened).toBe(true);
    unmount();
  });

  it('should set color cursors as moving on mouse down event', () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );
    fireEvent.mouseDown(container.querySelector('.lightness'), { button: 0 });
    expect(ref.current.handleMoving).toBe(true);
    expect(ref.current.handleType).toBe('lightness');
    unmount();
  });

  it('should not move color cursors as moving if clicked button is not 0 ' +
    '(left click)', () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );
    fireEvent.mouseDown(container.querySelector('.lightness'), { button: 1 });
    expect(ref.current.handleMoving).toBe(false);
    expect(ref.current.handleType).toBeNull();
    unmount();
  });

  it('should set color cursors as not moving on mouse up event', () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );
    fireEvent.mouseDown(container.querySelector('.lightness'), { button: 0 });
    expect(ref.current.handleMoving).toBe(true);
    expect(ref.current.handleType).toBe('lightness');
    fireEvent.mouseUp(document.body);
    expect(ref.current.handleMoving).toBe(false);
    unmount();
  });

  it('should fire onToggle event when opened/closed', async () => {
    const ref = createRef();
    const onToggle = jest.fn();
    const { container, unmount } = render(
      <ColorField ref={ref} onToggle={onToggle} />
    );

    await act(async () => { container.querySelector('input').focus(); });
    expect(ref.current.opened).toBe(true);
    expect(onToggle)
      .toHaveBeenCalledWith(expect.objectContaining({ opened: true }));

    fireEvent.click(document.body);
    expect(ref.current.opened).toBe(false);
    expect(onToggle)
      .toHaveBeenCalledWith(expect.objectContaining({ opened: false }));

    unmount();
  });

  it('should allow to reset input to its original value', async () => {
    const ref = createRef();
    const { container, unmount } = render(
      <ColorField ref={ref} value="#000" />
    );
    expect(ref.current.internalValue).toBe('#000');
    expect(ref.current.h).toBe(0);
    expect(ref.current.s).toBe(0);
    expect(ref.current.v).toBe(100);
    expect(ref.current.a).toBe(100);
    fireEvent
      .change(container.querySelector('input'), { target: { value: '#F00' } });
    expect(ref.current.internalValue).toBe('#F00');
    expect(ref.current.h).toBe(0);
    expect(ref.current.s).toBe(100);
    expect(ref.current.v).toBe(0);
    expect(ref.current.a).toBe(100);
    await act(async () => { ref.current.reset(); });
    expect(ref.current.internalValue).toBe('#000');
    expect(ref.current.h).toBe(0);
    expect(ref.current.s).toBe(0);
    expect(ref.current.v).toBe(100);
    expect(ref.current.a).toBe(100);
    unmount();
  });

  it('should allow to not have a preview', () => {
    const { container, unmount } = render(
      <ColorField previewEnabled={false} value="#F00" />
    );
    expect(container.querySelectorAll('.pigment').length).toBe(0);
    unmount();
  });
});
