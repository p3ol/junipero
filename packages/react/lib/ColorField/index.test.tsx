import { createRef } from 'react';
import {
  render,
  fireEvent,
  act,
  createEvent,
} from '@testing-library/react';

import ColorField, { ColorFieldRef } from '.';

describe('<ColorField />', () => {
  it('should render', () => {
    const { container, unmount } = render(<ColorField autoFocus={true} />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should update values when mouse is down and lightness handle is ' +
    'selected', async () => {
    const ref = createRef<ColorFieldRef>();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );
    fireEvent.mouseDown(container.querySelector('.lightness'), { button: 0 });
    expect(container).toMatchSnapshot('lightness handle moving');

    // Man testing is weird
    await act(async () => {
      window.pageXOffset = 50;
      window.pageYOffset = 50;

      // Simulate element width/height
      Object.defineProperty(
        ref.current.colorLightnessRef.current,
        'offsetWidth',
        { value: 100 }
      );
      Object.defineProperty(
        ref.current.colorLightnessRef.current,
        'offsetHeight',
        { value: 100 }
      );

      return Promise.resolve();
    });

    // Testing library (or jsdom?) doesn't support pageX/Y if passed
    // using fireEvent.mouseMove (lol), so we have to first create the event,
    // hard code the pageX and pageY values using .defineProperty, and then
    // give them straight to fireEvent (not fireEvent.mouseMove, this does
    // not work either)
    const moveEvent = createEvent
      .mouseMove(container.querySelector('.lightness'));
    Object.defineProperty(moveEvent, 'pageX', { value: 100 });
    Object.defineProperty(moveEvent, 'pageY', { value: 100 });

    fireEvent(container.querySelector('.lightness'), moveEvent);
    expect(container).toMatchSnapshot('color selected');
    unmount();
  });

  it('should update values when mouse is down and hue handle is ' +
    'selected', async () => {
    const ref = createRef<ColorFieldRef>();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );

    fireEvent.mouseDown(container.querySelector('.hue'), { button: 0 });
    expect(container).toMatchSnapshot('hue handle moving');

    await act(async () => {
      window.pageXOffset = 50;
      Object.defineProperty(
        ref.current.colorHueRef.current,
        'offsetWidth',
        { value: 100 }
      );

      return Promise.resolve();
    });

    const moveEvent = createEvent
      .mouseMove(container.querySelector('.hue'));
    Object.defineProperty(moveEvent, 'pageX', { value: 100 });
    fireEvent(container.querySelector('.hue'), moveEvent);

    expect(container).toMatchSnapshot('color selected');
    unmount();
  });

  it('should update values when mouse is down and alpha handle is ' +
    'selected', async () => {
    const ref = createRef<ColorFieldRef>();
    const { container, unmount } = render(
      <ColorField autoFocus={true} ref={ref} />
    );

    fireEvent.mouseDown(container.querySelector('.alpha'), { button: 0 });
    expect(container).toMatchSnapshot('alpha handle moving');

    await act(async () => {
      window.pageXOffset = 50;
      Object.defineProperty(
        ref.current.colorAlphaRef.current,
        'offsetWidth',
        { value: 100 }
      );

      return Promise.resolve();
    });

    const moveEvent = createEvent
      .mouseMove(container.querySelector('.alpha'));
    Object.defineProperty(moveEvent, 'pageX', { value: 100 });
    fireEvent(container.querySelector('.alpha'), moveEvent);
    expect(container).toMatchSnapshot('color selected');

    unmount();
  });
});
