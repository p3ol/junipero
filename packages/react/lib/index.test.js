import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
