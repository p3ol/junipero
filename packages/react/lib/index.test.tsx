import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
