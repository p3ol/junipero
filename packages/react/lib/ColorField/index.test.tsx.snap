// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ColorField /> > should render 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field pristine valid"
    id="junipero-dropdown-_r_1_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_1_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_2_"
      role="combobox"
    >
      <div
        class="junipero text-field pristine valid empty focused"
      >
        <input
          class="field"
          type="text"
          value=""
        />
      </div>
      <a
        class="pigment"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_1_-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(255, 0, 0);"
          >
            <a
              class="handle"
              style="transform: translate3d(0px, 0px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(255, 0, 0);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgb(255, 255, 255);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<ColorField /> > should update values when mouse is down and alpha handle is selected > alpha handle moving 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field dirty valid empty moving"
    id="junipero-dropdown-_r_d_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_d_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_e_"
      role="combobox"
    >
      <div
        class="junipero text-field dirty valid focused"
      >
        <input
          class="field"
          type="text"
          value="rgba(255, 255, 255, 0)"
        />
      </div>
      <a
        class="pigment"
        style="background-color: rgba(255, 255, 255, 0);"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_d_-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(255, 0, 0);"
          >
            <a
              class="handle"
              style="transform: translate3d(0px, 0px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(255, 0, 0);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgba(255, 255, 255, 0);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<ColorField /> > should update values when mouse is down and alpha handle is selected > color selected 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field dirty valid empty moving"
    id="junipero-dropdown-_r_d_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_d_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_e_"
      role="combobox"
    >
      <div
        class="junipero text-field dirty valid focused"
      >
        <input
          class="field"
          type="text"
          value="rgba(255, 255, 255, 0.5)"
        />
      </div>
      <a
        class="pigment"
        style="background-color: rgba(255, 255, 255, 0.5);"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_d_-menu"
      style="position: fixed; top: 10px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(255, 0, 0);"
          >
            <a
              class="handle"
              style="transform: translate3d(0px, 0px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(255, 0, 0);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(50px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgba(255, 255, 255, 0.5);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<ColorField /> > should update values when mouse is down and hue handle is selected > color selected 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field dirty valid empty moving"
    id="junipero-dropdown-_r_9_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_9_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_a_"
      role="combobox"
    >
      <div
        class="junipero text-field dirty valid focused"
      >
        <input
          class="field"
          type="text"
          value="#ffffff"
        />
      </div>
      <a
        class="pigment"
        style="background-color: rgb(255, 255, 255);"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_9_-menu"
      style="position: fixed; top: 10px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(0, 255, 255);"
          >
            <a
              class="handle"
              style="transform: translate3d(0px, 0px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(50px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(0, 255, 255);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgb(255, 255, 255);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<ColorField /> > should update values when mouse is down and hue handle is selected > hue handle moving 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field dirty valid empty moving"
    id="junipero-dropdown-_r_9_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_9_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_a_"
      role="combobox"
    >
      <div
        class="junipero text-field dirty valid focused"
      >
        <input
          class="field"
          type="text"
          value="#ffffff"
        />
      </div>
      <a
        class="pigment"
        style="background-color: rgb(255, 255, 255);"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_9_-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(255, 0, 0);"
          >
            <a
              class="handle"
              style="transform: translate3d(0px, 0px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(255, 0, 0);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgb(255, 255, 255);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<ColorField /> > should update values when mouse is down and lightness handle is selected > color selected 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field dirty valid empty moving"
    id="junipero-dropdown-_r_5_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_5_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_6_"
      role="combobox"
    >
      <div
        class="junipero text-field dirty valid focused"
      >
        <input
          class="field"
          type="text"
          value="#804040"
        />
      </div>
      <a
        class="pigment"
        style="background-color: rgb(128, 64, 64);"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_5_-menu"
      style="position: fixed; top: 10px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(255, 0, 0);"
          >
            <a
              class="handle"
              style="transform: translate3d(50px, 50px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(255, 0, 0);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgb(128, 64, 64);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<ColorField /> > should update values when mouse is down and lightness handle is selected > lightness handle moving 1`] = `
<div>
  <div
    class="junipero dropdown opened color-field dirty valid empty moving"
    id="junipero-dropdown-_r_5_"
  >
    <div
      aria-autocomplete="none"
      aria-controls="junipero-dropdown-_r_5_-menu"
      aria-expanded="true"
      aria-haspopup="listbox"
      class="inner dropdown-toggle opened"
      dir="ltr"
      id="junipero-dropdown-toggle-_r_6_"
      role="combobox"
    >
      <div
        class="junipero text-field dirty valid focused"
      >
        <input
          class="field"
          type="text"
          value="#000000"
        />
      </div>
      <a
        class="pigment"
        style="background-color: rgb(0, 0, 0);"
      />
    </div>
    <div
      class="junipero dropdown-menu color-menu"
      data-floating-ui-focusable=""
      id="junipero-dropdown-_r_5_-menu"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <ul
        class="menu-inner"
      >
        <div
          class="color-wheel"
        >
          <div
            class="lightness"
            style="background-color: rgb(255, 0, 0);"
          >
            <a
              class="handle"
              style="transform: translate3d(0px, 0px, 0);"
            />
          </div>
          <div
            class="controls"
          >
            <div
              class="sliders"
            >
              <div
                class="hue"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
              <div
                class="alpha"
                style="background-color: rgb(255, 0, 0);"
              >
                <a
                  class="handle"
                  style="transform: translate3d(0px, 0, 0);"
                />
              </div>
            </div>
            <div
              class="color-preview"
            >
              <div
                class="preview-inner"
                style="background-color: rgb(0, 0, 0);"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
